### PR TITLE
PERF: further improve take (reindex/unstack) for ArrayManager

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
A few related changes to `take_1d` / `ArrayManager.unstack`:

- Check in `ArrayManager.unstack` whether there will be missing values introduced or not, and if not, pass `allow_fill=False` to the `take` function, so this can take a more performant path (and doesn't need to check repeatedly for `indexer == -1`). And the same for `ArrayManager.reindex_indexer`.
- In the `take_1d`, if `allow_fill=False`, use the numpy `take` instead of our own cython implementation (numpy is a bit faster).
-  Move the `ensure_int64` check for the indexer out of the shared `_take_preprocess_indexer_and_fill_value`. This way, we can do this once in ArrayManager.unstack and avoid doing it repeatedly for the same indexer inside `take_1d`.

Case from the `frame_methods.py::Unstack.time_full_product` ASV benchmark:

```python
dtype = 'int'
m = 100
n = 1000

levels = np.arange(m)
index = pd.MultiIndex.from_product([levels] * 2)
columns = np.arange(n)
values = np.arange(m * m * n).reshape(m * m, n)

df = pd.DataFrame(values, index, columns)
df_am = df._as_manager("array")
```

with the combined changes, this gives:

```
In [2]: %timeit df_am.unstack()
1.09 s ± 68.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)  <-- master
276 ms ± 19.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)  <-- PR
```

### CODE REVIEW

File: pandas/core/array_algos/take.py
Comment (jorisvandenbossche) [inline pandas/core/array_algos/take.py:500]: Nothing changed in those lines of the code; it's only a change in indentation (which github does't seem to detect nicely). 

Only the first lines:

```
    if indexer is None:
        indexer = np.arange(arr.shape[axis], dtype=np.int64)
        dtype, fill_value = arr.dtype, arr.dtype.type()
    else:
        indexer = ensure_int64(indexer, copy=False)
```
have been moved out into `_take_nd_ndarray`

File: pandas/core/array_algos/take.py
Comment (jorisvandenbossche) [inline pandas/core/array_algos/take.py:89]: This part is copied from `_take_preprocess_indexer_and_fill_value`, so we can avoid doing that in `take_1d`

File: pandas/core/array_algos/take.py
Comment (jreback) [inline pandas/core/array_algos/take.py:138]: can you put asserts for these conditions

File: pandas/core/internals/array_manager.py
Comment (jreback) [inline pandas/core/internals/array_manager.py:1003]: why are you converting to int64? we use intp for indexing via take

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/array_manager.py
Comment (jorisvandenbossche) [reply pandas/core/internals/array_manager.py:1003]: That's existing code (just cut and paste). And our cython take algos always use int64:

https://github.com/pandas-dev/pandas/blob/d910a4435962505d3388d08a21e58bca6a9bf294/pandas/_libs/algos_take_helper.pxi.in#L69

File: pandas/core/array_algos/take.py
Comment (jorisvandenbossche) [reply pandas/core/array_algos/take.py:138]: Not really, as it's the point of this function to *not* do those asserts, but do them a level higher up in the calling code (and to be clear, this would just give a TypeError or AttributeError, it's not that it would segfault or so in case wrongly typed arguments are passed).

Comment (jreback) [reply pandas/core/array_algos/take.py:138]: sure you can we have asserts all of the place to enforce internal conditions, that's the point

File: pandas/core/internals/array_manager.py
Comment (jbrockmendel) [inline pandas/core/internals/array_manager.py:1007]: @jorisvandenbossche in looking at the benchmarks you posted in #39146 it seems like some of them involve optimizations implemented for AM that _could_ be implemented for BM but weren't.  is there anything here that falls into that category?

Comment (jorisvandenbossche) [reply pandas/core/internals/array_manager.py:1007]: > in looking at the benchmarks you posted in #39146 it seems like some of them involve optimizations implemented for AM that _could_ be implemented for BM but weren't

Are you thinking of something specific?

The optimization I do here above _can_ probably be useful if you have lots of non-consolidated blocks, but for the benchmark cases with single blocks posted in #39146, that wouldn't make a difference.

File: pandas/core/array_algos/take.py
Comment (jorisvandenbossche) [reply pandas/core/array_algos/take.py:138]: So the place where this function is called already has a `indexer = ensure_int64(indexer, copy=False)`. So adding a `assert indexer.dtype == "int64"` on the next line is then superfluous.

Comment (jreback) [reply pandas/core/array_algos/take.py:138]: pls do the asserts. documenting this is good, but an assert is way better.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/array_algos/take.py
Comment (jorisvandenbossche) [reply pandas/core/array_algos/take.py:138]: Have you read my response? 
The point of the changes in this PR is that we don't need to do those assertions 1000 times on the same data, but only once in the code calling this function. 

And the code calling this function asserts that the indexer is int64. And ensures that the input is an ndarray or EA.

File: pandas/core/array_algos/take.py
Comment (jbrockmendel) [inline pandas/core/array_algos/take.py:141]: entry for 'validate_indices has already been checked'?

File: pandas/core/array_algos/take.py
Comment (jbrockmendel) [inline pandas/core/array_algos/take.py:151]: good!  i was gonna mention i got segfaults when i tried using this elsewhere in this case

File: pandas/core/array_algos/take.py
Comment (jbrockmendel) [inline pandas/core/array_algos/take.py:155]: do we ever get to _take_preprocess_indexer_and_fill_value with allow_fill=False?

File: pandas/core/array_algos/take.py
Comment (jreback) [reply pandas/core/array_algos/take.py:138]: does this actually change the perf? asserts should not do that
if so that is a different issue

your comments don't change my mind here

File: pandas/core/internals/array_manager.py
Comment (jbrockmendel) [reply pandas/core/internals/array_manager.py:1003]: should they though?  i expect they ought to be intp_t

File: pandas/core/array_algos/take.py
Comment (jorisvandenbossche) [reply pandas/core/array_algos/take.py:141]: > entry for 'validate_indices has already been checked'?

Added (that's the same for `take_nd` as well though, so it's not a difference with take_nd)

File: pandas/core/array_algos/take.py
Comment (jorisvandenbossche) [reply pandas/core/array_algos/take.py:138]: Of course adding those asserts gives a performance overhead. Maybe not much for a single call, but the point is that we call this function *repeatedly* on partly the same input, and then this overhead adds up (eg in the unstack benchmark that I reference in the top post, the `take_1d` is called 100_000 times for a single unstack operation, while we can do those asserts a single time inside `unstack`). 
It's not necessarily the `assert` itself, but the actual check after the assert takes some time. For example, see the following call of `take_1d` using the current branch of this PR. With this PR, a single call (using data similarly as would be used in the referenced unstack operation) takes a bit more than 500ns:

```python
In [1]: from pandas.core.array_algos.take import take_1d

In [2]: from pandas.api.extensions import ExtensionArray

In [3]: arr = np.random.randn(10000)

In [4]: indexer = np.arange(100, dtype="int64")

In [5]: %timeit take_1d(arr, indexer, allow_fill=False)
561 ns ± 5.32 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [6]: %timeit assert isinstance(arr, (np.ndarray, ExtensionArray))
92.7 ns ± 1.01 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [7]: %timeit assert isinstance(indexer, np.ndarray)
70.1 ns ± 3.73 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [8]: %timeit assert indexer.dtype == "int64"
287 ns ± 1.72 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Doing all the assertions as I now documented them (ensuring it are actual arrays, ensuring the dtype of the indexer), would together give an almost 2x slowdown in this case. Only checking for ndarray would still give an almost 20% slowdown.

File: pandas/core/array_algos/take.py
Comment (jorisvandenbossche) [reply pandas/core/array_algos/take.py:162]: Yes, that's currently possible in `take_nd`, and a quick search reveals that we have a few cases throughout pandas that calls `take_nd` specifying this argument. That's probably something to optimize though.

File: pandas/core/array_algos/take.py
Comment (jorisvandenbossche) [reply pandas/core/array_algos/take.py:141]: (as another example, we don't actually check that indices are within bounds in take_nd/take_1d, which would actually silently give garbage data if passed wrong indexer, while a wrong dtype for the indexer even gives a normal error message)

Comment (jreback) [reply pandas/core/array_algos/take.py:141]: ok. note that maybe we should do release builds to avoid this issue generally. but i guess this is fine.

**Comment by jreback** [comment]
checks are failing (maybe just rebase needed)

File: pandas/core/array_algos/take.py
Comment (jbrockmendel) [reply pandas/core/array_algos/take.py:144]: thanks.  I _think_ calling take_1d without having done this check was the source of the segfaults i was getting

**Comment by jbrockmendel** [comment]
> it is likely this is a perf bottleneck (as we are moving between intp and int64 a fair amount)

arent these aliases on 64bit builds?

**Comment by jbrockmendel** [comment]
Can/should we just have take_nd do an earlier check for ndim == 1 and then dispatch to take_1d?

In NDArrayBackedExtensionArray.take if i change take_nd to

```
        if self.ndim == 1 and axis == 0 and allow_fill:
            indices = ensure_platform_int(indices)
            validate_indices(indices, len(self))
            new_data = take_1d(
                self._ndarray, indices, allow_fill=allow_fill, fill_value=fill_value
            )

        else:
             new_data = [what we do right now]
```

then on the time_full_product benchmark referenced https://github.com/pandas-dev/pandas/issues/39146#issuecomment-792817698 we shave off 25-30% relative to master (non-AM)

**Comment by jorisvandenbossche** [comment]
> Can/should we just have take_nd do an earlier check for ndim == 1 and then dispatch to take_1d?

That might be possible (although the check for EA and `np.asarray` might get a bit in the way for that). But we would still have the separate take_1d implemention, so I personally prefer to keep using that for explicitness in those places where we know it's only about 1D arrays (like the places in ArrayManager where I am using it now).

**Comment by jbrockmendel** [comment]
> so I personally prefer to keep using that for explicitness in those places where we know it's only about 1D arrays (like the places in ArrayManager where I am using it now).

im fine with that, just suggesting if we have a clear optimization for the general `take` lets use it

**Comment by jorisvandenbossche** [comment]
> im fine with that, just suggesting if we have a clear optimization for the general take lets use it

Certainly, see https://github.com/pandas-dev/pandas/pull/40405

**Comment by jorisvandenbossche** [comment]
For the rest, any remaining comments before merging this?

**Comment by jbrockmendel** [comment]
ill take another look this afternoon

File: pandas/core/internals/array_manager.py
Comment (jbrockmendel) [inline pandas/core/internals/array_manager.py:1102]: can this chunk of code be made to match Block._unstack (or vice-versa) more closely?

e.g. on L1094 you check unstacker.mask.all() which I think is related to the comment in Block._unstack `# TODO: in all tests we have mask.all(); can we rely on that?` but bc `mask` is defined there differently it is not obvious

File: pandas/core/internals/array_manager.py
Comment (jbrockmendel) [reply pandas/core/internals/array_manager.py:1009]: im increasingly confident that intp_t is the correct thing to do, but its out of scope for this PR, xref #40390

File: pandas/core/internals/array_manager.py
Comment (jorisvandenbossche) [reply pandas/core/internals/array_manager.py:1102]: Assuming you are referring to the EA-specific `Block.unstack` (as the other one has even less in common): that's for a single column while this is for multiple columns at once, and needs to deal with creating blocks, with placement, etc. So personally I don't see much possibilities to share code. 

> e.g. on L1094 you check unstacker.mask.all() which I think is related to the comment in Block._unstack

As far as I understand the Block version, I don't think that's related. For Block/BlockManager.unstack, that `mask` value somehow indicates which of the columns need to end up in the result (I don't really understand that code, though, as in the ArrayManager I am simply using the result of `unstacker.get_new_columns` without any filtering of that afterwards), while here I only check the `mask` to potentially use a faster `take` for performance-reasons, but that doesn't influence the end result.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm @jbrockmendel merge when ready

File: pandas/core/internals/array_manager.py
Comment (jbrockmendel) [reply pandas/core/internals/array_manager.py:1102]: > I don't think that's related

OK.

>  (I don't really understand that code, though

at some point we should make sure that one of us (active maintainers, not just the two of us) have a handle on this.

**Comment by jbrockmendel** [comment]
LGTM.  Merging this evening absent other comments.

**Comment by jreback** [comment]
thanks @jorisvandenbossche

**Pull request merged**
